### PR TITLE
Lp1777667 Quantize with different tempos 

### PR DESCRIPTION
**Pull request by daschuer**
This is a proof of concept branch for a new quantize strategy. 
It was build on top of https://github.com/mixxxdj/mixxx/pull/1715
https://bugs.launchpad.net/mixxx/+bug/1777667

If we agree this is a good way to fix the issue, this needs to be some love to streamline the code.

### CODE REVIEW

**Comment by kazakore** [comment]
I've been using this PR as my main goto version for a while as the issue fixed by #1715  and the issue fixed by this PR are both quite important to me. What are the chances of the flagged issues being resolved and these fixes making it into 2.1.2?

**Comment by daschuer** [comment]
I agree with you here, we need a second team member opinion.
@ywwg What do you think?

**Comment by Be-ing** [comment]
IMO we should be much more conservative about what goes into bugfix releases. I think this should go into master.

**Comment by kazakore** [comment]
Yeah I would be happy enough with this one in Master, #1715 also looks to be stuck (and semi-related) and is much too serious to wait for 2.2 really though IMO....

**Comment by daschuer** [comment]
If this is considered as ok and not breaking anything,  it can IMHO go to any version. 
@kazakore did you see anything suspicious?  
The old way of sync phase is realy useless. @Be-ing: try a 60 bpm and a 120 bpm track.

**Comment by kazakore** [comment]
It worked perfectly as I would expect with anything I've thrown at it personally (even with my most extreme case of 80BPM to 300BPM, which unfortunately Sync doesn't work with, doubled BPM of everything below 80 recently and might actually bump that to a higher value as well.) 

Means you now have Quantise on an if you're bringing in a different tempo track drop single beat snippets from the next tune actually on the beat perfectly, rather than them being at some weird offset point. Good if you want to mix a bit of non-syncopated live music into something electronic with a good grid etc.

**Comment by Be-ing** [comment]
We have already been burned by rushing substantial engine changes into 2.1.1 without beta testing. The issue this fixes indeed should be fixed, but it is far from critical.

**Comment by kazakore** [comment]
But I also agree with Be's logic, hence I don't feel it's my place to say what level of bug fix should go in a point release. All I can say is I consider this fix to be a success :-)

**Comment by kazakore** [comment]
Why was this put in the 2.3 Milestone and not 2.2?

**Comment by daschuer** [comment]
Because we have not tested enough. It this one itching you too much?

**Comment by kazakore** [comment]
Yes because it effects using decks C & D as sample triggers. If you mix using decks A and B but have a 4 deck controller you can load tracks or samples to decks C and D and just use the HotCue triggers for playing samples such as stabs and vocal snippets. Without this patch this trigger completely out of time if you have Quantise enabled. It's rather a glaring fault IMPO. 

I guess it also effects the actual Samplers but as I mostly use hotcues within the full track, rather than actual edited samples, I can't say I've tested that...

**Comment by kazakore** [comment]
Still desperate for this fix! My self-built version definitely has some other minor bugs in it and it's a shame to be missing some of the other newer features, but I couldn't get this to Merge with 2.1 or 2.2 stable so had to go with it onto #1745 (as that is what I had used before which does merge and was a fix that stopped my system glitching on audio at times.)

Biggest reason I'm finding I need it desperately is I'm often playing hip-hop, reggae and jungle in the same set. Jungle is double the tempo of the other two genres (and I have so many tracks I want to keep them listed as real tempo.) After a few drinks I've been finding I do have to turn on and rely on Quantise (plus my laptop doesn't have the best specs for low latency) and without this fix tracks at double or half tempo do not trigger at the right time. Which means at least every other mix, if not more often a lot of the time, I can not use hot cues to jump and trigger section while the track is playing.

It looks like somebody has had at least a brief look since I mentioned it in Zulip a little while ago. Here's crossing my fingers it get successfully merged somewhere soon, even if only in Master....

**Comment by kazakore** [comment]
I've been using a fairly recent build of Master lately (r6835) and have noticed something strange with it. Seemingly this issue only presents itself when my controller isn't connected, when I have my controller plugged it the hotcue triggering quantises correctly whether I use the pads on the controller or the keys on the keyboard. If no controller is tested and I try triggering with the keyboard then triggering does not get quantised correctly.

As such this makes having this patch added to Master less of an issue for me personally (as obviously I use the controller when performing) but it does seem to imply something strange is going on as connecting a controller should not make such a difference...

**Comment by daschuer** [comment]
So you use this patch now for a while without any issues? @Be-ing any objections to merge it now? 

> have noticed something strange with it

@kazakore please explain this a bit more. Is this PR really a solution? I am in doubt, because this is not controller related. Please explain a bit more what should happen and what "strange" happens instead.

**Comment by kazakore** [comment]
I'll have to test 2.2 later to confirm but 100% definitely with 2.1 when quantise is enabled and you try mixing tracks of different tempo (most often for me something of double or half the tempo but may sometimes be sample triggering of random difference) then triggering the hotcues does not align with beatmarkers of the two tracks (with double tempo being triggered it lets you trigger at the half way point as well, which is good) but they get shifted either early or late but not by the amount they would be if Quantise was off.

With this patch they always trigger correctly on the beatmarkers, but it was built when there were other minor issues (eg waveforms aren't as smooth as a later patch and some other things I can't remember now.) Used it live a few times with no issues for what it was designed to solve, also used it when organising tracks with no controller connected and thus can say for 100% it quantises correctly with and without the controller connected.

I liked the idea of the new In/Out Markers so I could set a visible cue of where not to play certain tracks past so used that for a while. On initial testing with no controller connected (IE when I arrange my tracks, set markers etc) I tested and this issue was still present. Use the Master r6835 build to make a mix and triggering is then correctly quantised. Tested a few times to make sure it wasn't a fluke and definitely seems to work when the controller is plugged in and not when there is no controller on that version of Master. Built the latest version yesterday but am as yet to get to test it....

So it seems at some point between this PR and the r6835 and patch has been added which seems to fix the quantisation but only when a controller is connected. Fixing only under these circumstances is what I think is weird, not any particular behaviour. I think working out what that may be would be better than just merging this blind, as it does now seem to work under performance (controller connected) conditions.

**Comment by daschuer** [comment]
Interesting, normally the controller and the Gui are using the same controls and the same syncing code. So there seems to be something special in you case. Which controller mapping do you use?
Please describe step by step what you do, on your controller to trigger the issue. What do you do step by step to reproduce the issue with the Gui.

**Comment by kazakore** [comment]
Thanks for prompting me to test more. My above comment is incorrect on what causes the difference, I originally couldn't repeat it and wondered what was going on. Sorry for the red herring.

What actually makes the difference is if Master Sync is enabled on the decks. With it on then Quantising is correctly performed on the beatgrid (including half beat when playing a double tempo track :) ) but without the Master Sync Quantise is off (it appears to shift about 50% in the right direction guessing by eye. 

I don't think I had used the Master Sync much at all until recently so just tested and this is in fact the same in 2.1.

I thought the Always In Beat Sync aspect of Master Sync should only really affect tracks with variable tempo or possibly things like keeping beats locked if you do a spinback etc. Quantise on its own should do it for Play, HotCues, Reverse etc but it doesn't.

Also a note on the double/half beats triggering. Strangely if anything this seems to be backwards (although I would prefer it works both ways. If you have say a 160bpm track playing and bring in a 80bpm track you can only trigger the Cues every alternate beatmarker of the playing track. Sometimes this is on the offbeat as Mixxx doesn't always have the downbeat set correctly (really annoying!) If you mix a160bpm track into an 80bpm track then you can trigger at the point half way between the playing tracks beatgrid.

**Comment by daschuer** [comment]
> What actually makes the difference is if Master Sync is enabled on the decks. With it on then Quantising is correctly performed on the beatgrid (including half beat when playing a double tempo track :) ) but without the Master Sync then Quantising is off (it appears to shift about 50% in the right direction guessing by eye.)

Is Master sync = both sync buttons latched?
I will have a look.

**Comment by kazakore** [comment]
>Is Master sync = both sync buttons latched?
>I will have a look.

Yeah that's what I meant, sorry for using incorrect terminology.

**Comment by daschuer** [comment]
This is mergeable again.

File: src/engine/controls/bpmcontrol.cpp [hunk 1]
```diff
@@ -725,6 +726,160 @@ double BpmControl::getNearestPositionInPhase(
...
 }
 
+double BpmControl::getBeatMatchPosition(
+        double dThisPosition, bool respectLoops, bool playing) {
+    // Without a beatgrid, we don't know the phase offset.
+    if (!m_pBeats) {
+        return dThisPosition;
+    }
+    // Master buffer is always in sync!
+    if (getSyncMode() == SYNC_MASTER) {
```
Comment (ywwg) [inline src/engine/controls/bpmcontrol.cpp]: There's one case where this isn't true: if this deck has master sync on and the other deck does not, and is playing.  When we press play, we expect the new deck to sync phase to the other non-sync deck. Luckily, the fix for this should just be to remove this clause.

Comment (ywwg) [reply src/engine/controls/bpmcontrol.cpp]: (It looks like this code is copied from existing code, check out my mastersync-onemaster branch to see how it changed)

**Review by ywwg — CHANGES_REQUESTED** [review]
This does appear to fix lp#1799080 so I am interested in seeing this finished.

**Comment by kazakore** [comment]
Phase syncing should only done on Quantise being enabled, whether sync is locked or not should make zero difference.

**Comment by ywwg** [comment]
> Phase syncing should only done on Quantise being enabled, whether sync is locked or not should make zero difference.

I'm not sure what you're responding to specifically, but yes, this is correct.  My note above is about the case where a deck can be in SYNC_MASTER mode and it definitely should sync phase.

**Comment by daschuer** [comment]
Cool, so because of https://bugs.launchpad.net/mixxx/+bug/1799080 we should consider to move this to the 2.3.0 milestone.

File: src/engine/controls/bpmcontrol.cpp (hunk 1 above)
Comment (daschuer) [reply src/engine/controls/bpmcontrol.cpp]: I am afraid that just removing this clause will break other edge cases. 
If the track is already playing and is SYNC_Master, than this condition is probably still desired. 
I will check.

**Comment by daschuer** [comment]
Tests are failing,

**Comment by ywwg** [comment]
> Cool, so because of https://bugs.launchpad.net/mixxx/+bug/1799080 we should consider to move this to the 2.3.0 milestone

Together with #2376 we can make some major fixes to beatmap mixing.  I don't know what our timeline is or how close we are to freeze though.

**Comment by Holzhaus** [comment]
I thought we are already in a kind of freeze. As far as I'm concerned we are only delaying 2.3 to include PR #2345. Since there's still ongoing discussion on that PR and this one fixes bug, I think we could justify merging this. @Be-ing @uklotzde What do you think?

**Comment by ywwg** [comment]
I don't think either this or my PR are worth breaking a freeze.  Let's aim for 2.4 to get beatmaps working once and for all

**Comment by ywwg** [comment]
I think this PR has issues when only one track is loaded.  When I have one beatmap track loaded and sync is off and I push play, it skips ahead and starts playing at the second beat.

**Comment by daschuer** [comment]
I cannot reproduce this after the latest changes. 
Can you re-test?

**Comment by ywwg** [comment]
Ah, you've made some changes.  I notice that some of the tests are failing now:

`
[  FAILED  ] EngineSyncTest.HalfDoubleBpmTest
[  FAILED  ] EngineSyncTest.SyncPhaseToPlayingNonSyncDeck
[  FAILED  ] EngineSyncTest.QuantizeImpliesSyncPhase
[  FAILED  ] EngineSyncTest.SeekStayInPhase
[  FAILED  ] EngineSyncTest.QuantizeHotCueActivate
`

**Comment by ywwg** [comment]
tests are passing for me now, nevermind

File: src/engine/controls/bpmcontrol.cpp [hunk 2]
```diff
@@ -725,6 +726,162 @@ double BpmControl::getNearestPositionInPhase(
...
+    BeatsPointer otherBeats = otherTrack ? otherTrack->getBeats() : BeatsPointer();
+
+    // If either track does not have beats, then we can't adjust the phase.
+    if (!otherBeats) {
+        return dThisPosition;
+    }
+
+    double dOtherPosition = pOtherEngineBuffer->getExactPlayPos();
+
+    if (!BpmControl::getBeatContext(otherBeats, dOtherPosition,
```
Comment (ywwg) [inline src/engine/controls/bpmcontrol.cpp]: I don't think this code correctly handles half/double bpm sync situations.  In trying to merge this with my one mastersync-onemaster code, it picks the wrong beats to sync against.

**Comment by ywwg** [comment]
OK I finally figured out a reproducible issue: when pressing play, the track tries to sync phase even if there are no other tracks loaded.  This exercises the line "pOtherEngineBuffer = getEngineBuffer();" in getNearestPositionInPhase or getBeatMatchPosition. This can cause a problem if the current playback position is before the first beat of a beatmap track.  The engine tries to seek to the same beat fraction as itself, but because the track is beatmapped, it can't find a location before the first beat.  So instead it skips to the same beat fraction of the next beat, and the track skips ahead by a beat.

I think we can solve this by taking out the code where seek phase tries to sync to itself (we have to be careful, I think this code exists for looping)

**Comment by daschuer** [comment]
What is the best way to continue here, I think it makes sense to test and review a merged branch with your and my PR together. Do you already have one? 
Maybe we should create a mixxxdj/mixxx branch and issue future PRs against that.
What do you think?  

I think some more testing and refactoring is required. I cannot point it out exactly yet, but i have the feeling that there is useless code left over, that obscured a clear view. Maybe we can push it a bit in the CO snapshot direction.

**Comment by ywwg** [comment]
> What is the best way to continue here, I think it makes sense to test and review a merged branch with your and my PR together. Do you already have one?
> Maybe we should create a mixxxdj/mixxx branch and issue future PRs against that.
> What do you think?

sorry I didn't see this message before.  Yes I think we should merge our two branches together. Maybe what we should do is branch current master for 2.3 and then start working on releasing it, than merge these into master.

As for clarifying master sync and the three steps you mention, I think that's a good idea but I want to start with a solid base to refactor from.  I think our two PRs fix a lot of issues that are a problem right now, and also clean things up incrementally.  I would prefer to continue this approach with incremental cleanup and new tests in master rather than a huge branch developed off to the side. At every stage of development we should be able to make things a little better and a little more stable.  I believe if we set a grandiose goal of "fixing the engine completely" before merging, it will never happen. The incremental approach has worked much better for us.

**Comment by ywwg** [comment]
I did attempt to merge our branches together, and mostly it worked fine. I had an issue where I'm not sure that getBeatMatchPosition is even necessary. getNearestPositionInPhase works fine for me in all cases.  Why did you need to add it?

Other than that, and the one bug I mentioned (the longstanding problem of beat maps not having pre-first-beat markers), it worked.

**Comment by daschuer** [comment]
getBeatMatchPosition() ensures that even if the tracks have no synced tempos that the beats around the playposition are played in sync.  getNearestPositionInPhase sync the beat fraction, which does not make any sense in case of  different tempos. 

Thinking about it, I think there is still a bug, because is also syncs the past beat if it is closer. I will fix it. 

Would you support merging both branches to master before 2.3. I am slightly in favor of this to not create an other loose end.

**Comment by ywwg** [comment]
(fyi I am traveling so I'm slow to reply).  Yes I'd be in favor of merging these sooner because they fix some real bugs that make DJing beatmapped tracks extremely difficult.

**Comment by ywwg** [comment]
I'm back!  How about we merge this first, and then I'll update mine and merge it second?

**Comment by daschuer** [comment]
Yes this will work. Unfortunately test are failing now. I hope I can fix it tonight.

**Comment by daschuer** [comment]
@ywwg: here is my current state. 
There is probably still an issue that a seek does not update all required variables.
Normally my last commit should fix this but it is still failing. 
I need to wrap my head around it once I have time.
In the meantime it would be helpful if you have an eye on that code.
Thank you.

**Comment by ywwg** [comment]
>  In the meantime it would be helpful if you have an eye on that code.

ok I'll take a look, thanks!

File: src/track/beatmap.cpp [hunk 3]
```diff
@@ -490,36 +490,6 @@ void BeatMap::removeBeat(double dBeatSample) {
     emit updated();
 }
 
-void BeatMap::moveBeat(double dBeatSample, double dNewBeatSample) {
```
Comment (ywwg) [inline src/track/beatmap.cpp]: Are these changes to BeatMap intended?  This feels like something left over from a merge.

File: src/engine/controls/bpmcontrol.cpp [hunk 4]
```diff
@@ -725,6 +726,142 @@ double BpmControl::getNearestPositionInPhase(
...
+
+            // Syncing to after the loop end.
+            if (end_delta > 0 && loop_length > 0.0) {
+                int i = end_delta / loop_length;
+                dNewPlaypos = loop_start_position + end_delta - i * loop_length;
+
+                // Move new position after loop jump into phase as well.
+                // This is a recursive call, called only twice because of
+                // respectLoops = false
+                dNewPlaypos = getNearestPositionInPhase(dNewPlaypos, false, playing);
```
Comment (ywwg) [inline src/engine/controls/bpmcontrol.cpp]: do you mean to recurse into getnearestpositioninphase, or should this be getbeatmatchposition?

**Comment by daschuer** [comment]
@ywwg OK, now it is working with a 1/8 Threshold to match past beats. 
Pleas test again.

**Comment by ywwg** [comment]
This is working now!

File: src/engine/controls/bpmcontrol.cpp [hunk 5]
```diff
@@ -803,14 +803,29 @@ double BpmControl::getBeatMatchPosition(
     double thisDivSec = (dThisNextBeat - dThisPosition)
             / dThisSampleRate / dThisRateRatio;
 
+    if (dOtherBeatFraction < 1.0/8) {
```
Comment (ywwg) [inline src/engine/controls/bpmcontrol.cpp]: I was thinking an eighth note, not an eighth of a beat (which is a quarter note).  An eighth note would be 50% of a beat, which we decided is too long, but I think this is too quick.  Can you make it 1/4 instead?

**Comment by ywwg** [comment]
I still see one test failure:
`src/test/enginesynctest.cpp:1606: Failure
Expected equality of these values:
  ControlObject::get(ConfigKey(m_sGroup1, "beat_distance"))
    Which is: 0.15092970521541954
  ControlObject::get(ConfigKey(m_sGroup2, "beat_distance"))
    Which is: 0.11609977324263039
in ~EngineMaster()
`

**Comment by ywwg** [comment]
we'll merge this in and then I'll see if I can merge it with my onemaster change.

**Comment by daschuer** [comment]
> I still see one test failure:

Strange, I do not see it and the CI does not as well. 
Does it always fail?

File: src/engine/controls/bpmcontrol.cpp (hunk 5 above)
Comment (ywwg) [reply src/engine/controls/bpmcontrol.cpp]: I am thinking about it musically, and a sixteenth note feels like a good amount of time to me.  The current value feels a little too strict

**Comment by ywwg** [comment]
I reset your branch and tried again and it works.  I think I might have had a diff somewhere.  Can you remove your debug statements?

**Comment by daschuer** [comment]
Done.

**Comment by daschuer** [comment]
Regarding the 1/8 beat distance.
What is your test case?
Mine is playing a 120 Bpm track with a 125 Bpm track.
I cue the 120 Bpm track at a downbeat.
Than hit play when the first track plays a downbeat as well.
Right in time I hear both beats together and the second beat as a double beat.
When I am to late > 1/32 note, I hear the downbeat of the first track and the second beat played together. A good thing IMHO.

When I increat the theashold, and I am too late but within the theashold, I hear the downbeat of the new track and a double beat. 
That sounds IMHO not as good as before when the second beat matches.

What do you think, what is your use case?

**Comment by ywwg** [comment]
I'm ok keeping it at a 32nd note for now.  I can follow up later if I find enough examples where it's an issue.

**Comment by ywwg** [comment]
LGTM!

**Comment by daschuer** [comment]
I think we should merge it now. This fixes an issue that sync code uses outdated data after seek.
After the additional review and test round it should be mature for beta phase. 
Else the is a risk for code rod. 
@Be-ing: What do you think?

**Comment by ywwg** [comment]
Targetting to 2.3 seems right to me because this fixes a longstanding bug.  Any objections?

**Comment by rryan** [comment]
> I think we should merge it now. This fixes an issue that sync code uses outdated data after seek.
> After the additional review and test round it should be mature for beta phase.

One less 2018 PR on the books would be nice. I think you should merge if @ywwg has LGTM'd and we're not in code freeze yet.
